geniza/corpus/metadata_export.py

```diff
@@ -0,0 +1,221 @@
+    "collection",
+]
+
+import csv
+import os
+
+from django.conf import settings
+from django.contrib.postgres.aggregates import ArrayAgg
+from django.contrib.sites.models import Site
+from django.db.models.query import Prefetch
+from django.http import StreamingHttpResponse
+from django.utils import timezone
+from rich.progress import track
+
+script_user = settings.SCRIPT_USERNAME
+site_domain = Site.objects.get_current().domain.rstrip("/")
+url_scheme = "https://"
+
+from geniza.common.utils import Echo, timeprint
+from geniza.corpus.models import Document, TextBlock
+
+
+class Exporter(object):
+    model = None
+    str_plural = "objects"
+    csv_fields = []
+
+    def __init__(
+        self, qset=None, progress=CSV_EXPORT_PROGRESS, sep_within_cells=SEP_WITHIN_CELLS
+    ):
+        self.qset = qset
+        self.progress = progress
+        self.sep_within_cells = sep_within_cells
+
+    def csv_filename(self):
+        return (
+            f'geniza-{self.str_plural}-{timezone.now().strftime("%Y%m%dT%H%M%S")}.csv'
+        )
+
+    def get_queryset(self):
+        return self.model.objects.metadata_prefetch() if not self.qset else self.qset
+
+    def get_export_data_dict(self, obj):
+        # @NOTE: THIS NEEDS TO BE SUBCLASSED
+        pass
+
+    def iter_export_data_as_dicts(self):
+        timeprint("iter_export_data_as_dicts")
+        # get qset
+        qset = self.get_queryset()
+
+        # progress bar?
+        iterr = (
+            qset
+            if not self.progress
+            else track(qset, description=f"Writing rows to file")
+        )
+
+        # save
+        yield from (self.get_export_data_dict(obj) for obj in iterr)
+
+    def iter_export_data_as_csv(self, fn=None, pseudo_buffer=False):
+        timeprint("iter_export_data_as_csv")
+        with (
+            open(self.csv_filename() if not fn else fn, "w")
+            if not pseudo_buffer
+            else Echo()
+        ) as of:
+            writer = csv.DictWriter(
+                of, fieldnames=self.csv_fields, extrasaction="ignore"
+            )
+            yield writer.writeheader()
+            yield from (
+                writer.writerow(docd) for docd in self.iter_export_data_as_dicts()
+            )
+
+    def write_export_data_csv(self, fn=None):
+        timeprint("write_export_data_csv")
+        if not fn:
+            fn = self.csv_filename()
+        for row in self.iter_export_data_as_csv(fn=fn, pseudo_buffer=False):
+            pass
+
+    def http_export_data_csv(self, fn=None):
+        timeprint("http_export_data_csv")
+        if not fn:
+            fn = self.csv_filename()
+        iterr = self.iter_export_data_as_csv(pseudo_buffer=True)
+        response = StreamingHttpResponse(iterr, content_type="text/csv; charset=utf-8")
+        response["Content-Disposition"] = f"attachment; filename={fn}"
+        return response
+
+
+class DocumentExporter(Exporter):
+    model = Document
+    str_plural = "documents"
```

Comment:
should be able to get this from the plural verbose name on the model: `Documents._meta.verbose_name_plural`

https://docs.djangoproject.com/en/4.1/ref/models/options/#verbose-name-plural